controller/rest/rest.go

```diff
@@ -1806,6 +1877,22 @@ func StartRESTServer(isNewCluster bool, isLead bool) {
 		PreferServerCipherSuites: true,
 		CipherSuites:             utils.GetSupportedTLSCipherSuites(),
 	}
+
+	// tlsCertificate is only generated when default location has no files
+	// so we can check if tlsCertificate is nil to see if we should adopt dynamically generated certificate.
+	certFileName := defaultSSLCertFile
+	keyFileName := defaultSSLKeyFile
+	if tlsCertificate != nil {
+		// When provide certificate from memory, certFileName and keyFileName have to be empty strings.
+		certFileName = ""
+		keyFileName = ""
+		config.GetCertificate = func(*tls.ClientHelloInfo) (*tls.Certificate, error) {
+			tlsMutex.RLock()
+			defer tlsMutex.RUnlock()
+			return tlsCertificate, nil
```

Comment:
if we use Rlock here, is it guarantee this is the latest certificate?

Reply:
If I understand your question correctly, you mean if there is a possibility that `GetCertificate()` would get old certificate once new one is generated?  

So while RLock() doesn't block RLock(),  the scope protected by RLock() and the scope protected by Lock() will be mutual exclusive.  That means once the certificate is created and assigned to `tlsCertificate` in `NotifyNewCert()`, it will then be used by the `GetCertificate()` function when handshake happens next time.

Reply:
I see, thanks for the context.